### PR TITLE
Block

### PR DESCRIPTION
**Pull request by turuslan**
### Description of the Change
Add `BlockHeader` struct with cbor encode/decode.
Add cbor encode/decode methods for optional and vector.

### Benefits
`BlockHeader` is required for tipset and chain

### Possible Drawbacks 
None

### CODE REVIEW

File: core/codec/cbor/cbor_decode_stream.hpp [hunk 1]
```diff
@@ -59,6 +59,34 @@ namespace fc::codec::cbor {
       return *this;
     }
 
+    template <typename T>
+    CborDecodeStream &operator>>(boost::optional<T> &optional) {
+      if (isNull()) {
+        optional = boost::none;
+        next();
+      } else {
+        T value;
```
Comment (masterjedy) [inline core/codec/cbor/cbor_decode_stream.hpp]:
```suggestion
        T value{};
```

File: core/codec/cbor/cbor_decode_stream.hpp [hunk 2]
```diff
@@ -59,6 +59,34 @@ namespace fc::codec::cbor {
...
+
+    /// Decodes list elements into vector
+    template <typename T>
+    CborDecodeStream &operator>>(std::vector<T> &values) {
+      auto n = listLength();
+      auto l = list();
+      values.clear();
+      values.reserve(n);
+      for (auto i = 0u; i < n; ++i) {
+        T value;
```
Comment (masterjedy) [inline core/codec/cbor/cbor_decode_stream.hpp]:
```suggestion
        T value{};
```

File: core/primitives/block/block.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include "primitives/ticket/ticket_codec.hpp"
+
+namespace fc::primitives::block {
+  using primitives::BigInt;
+  using primitives::address::Address;
+  using primitives::ticket::EPostProof;
+  using primitives::ticket::Ticket;
+
+  // TODO(turuslan): FIL-72 signature
+  using Signature = std::vector<uint8_t>;
```
Comment (masterjedy) [inline core/primitives/block/block.hpp]: is it a universal signature type? not pls?

File: test/core/codec/cbor/cbor_test.cpp [hunk 4]
```diff
@@ -73,6 +73,21 @@ TEST(Cbor, Null) {
...
+/** Optional CBOR encoding and decoding */
+TEST(Cbor, Optional) {
+  boost::optional<int> empty;
+  EXPECT_OUTCOME_EQ(encode(empty), "F6"_unhex);
+  EXPECT_OUTCOME_EQ(decode<boost::optional<int>>("F6"_unhex), empty);
+  EXPECT_OUTCOME_EQ(encode(boost::make_optional(3)), "03"_unhex);
+  EXPECT_OUTCOME_EQ(decode<boost::optional<int>>("03"_unhex), 3);
+}
+
+TEST(Cbor, Vector) {
```
Comment (masterjedy) [inline test/core/codec/cbor/cbor_test.cpp]: add comment

File: test/core/primitives/block/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright Soramitsu Co., Ltd. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+addtest(block_test
+    block_test.cpp
+    )
+target_link_libraries(block_test
+    address
```
Comment (masterjedy) [inline test/core/primitives/block/CMakeLists.txt]: tests should be linked with as little libraries number as possible. ideally, with only one.
if libraries are required for the library being tested, they should be linked to it, not to the test.
I mean address, cbor and tockets libraries.

Comment (masterjedy) [reply test/core/primitives/block/CMakeLists.txt]: I see, there is no block library. maybe add such library and link all the dependencies to it>

File: core/primitives/block/block.hpp (hunk 3 above)
Comment (turuslan) [reply core/primitives/block/block.hpp]: Lotus sources suggest so

File: test/core/primitives/block/block_test.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "primitives/block/block.hpp"
+
+#include <gtest/gtest.h>
+#include "testutil/cbor.hpp"
+
+/// Block header CBOR encoding and decoding
```
Comment (Harrm) [inline test/core/primitives/block/block_test.cpp]: Please comment in given-when-then format, presumably splitting on a few tests.

File: core/primitives/block/block.cpp [hunk 7]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "primitives/block/block.hpp"
```
Comment (Harrm) [inline core/primitives/block/block.cpp]: Is it needed?

File: test/core/codec/cbor/cbor_test.cpp [hunk 8]
```diff
@@ -73,6 +73,22 @@ TEST(Cbor, Null) {
   EXPECT_FALSE(CborDecodeStream("01"_unhex).isNull());
 }
 
+/** Optional CBOR encoding and decoding */
```
Comment (Harrm) [inline test/core/codec/cbor/cbor_test.cpp]: And here given-when-then, too, please. It just helps to write more meaningful tests, for edge cases for example.

File: core/primitives/block/block.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+  struct BlockHeader {
```
Comment (igor-egorov) [inline core/primitives/block/block.hpp]: The initializer for every struct field looks like a workaround.

File: core/primitives/block/block.cpp (hunk 7 above)
Comment (turuslan) [reply core/primitives/block/block.cpp]: [> Pls consider adding a library for block linking all the dependencies to it, and other minor issues.](https://github.com/filecoin-project/cpp-filecoin/pull/68#pullrequestreview-343509730)
